### PR TITLE
Revise entropy check to work without floating IP

### LINKED ISSUE #650: Revise entropy check to work without floating IP
Floating IPs are not viable for every provider. It might be possible to do without them, as follows:

- include necessary commands in cloud-init, see https://github.com/osism/openstack-simple-stress/blob/main/openstack_simple_stress/main.py#L275
- parse console output as provided by Nova, see https://github.com/osism/openstack-simple-stress/blob/main/openstack_simple_stress/main.py#L208

### PR DESCRIPTION
**Pull request by mbuechse**
resolves #650

### CODE REVIEW

**Review by martinmo — CHANGES_REQUESTED** [review]
Nice feature, and nice to have less dependencies listed in `requirements.txt` :+1:

Left some remarks above, one of it related to important functionality (timeout).

Function `retry()` isn't referenced anymore and can be removed.

File: Tests/iaas/entropy/entropy-check.py [hunk 1]
```diff
@@ -458,13 +384,32 @@ def print_result(check_id, passed):
...
+    for line in lines:
+        idx = line.find("_scs-test-")
+        if idx != -1:
+            section = line[idx + 10:].split()[0]
+            if section == 'end':
+                section = None
+            indent = idx
+            continue
+        if section:
+            collected[section].append(line[indent:])
```
[the remark this answers is not in the document]
Comment (mbuechse) [reply Tests/iaas/entropy/entropy-check.py]: ae425f2b1b10266306322787056fd0183c04511a

File: Tests/iaas/entropy/entropy-check.py [hunk 2]
```diff
@@ -518,27 +463,23 @@ def main(argv):
...
-                            retry(fconn.open, exc_type="NoValidConnectionsError,TimeoutError")
-                            # virtio-rng is not an official test case according to testing notes,
-                            # but for some reason we check it nonetheless (call it informative)
-                            check_virtio_rng(fconn, image, server.flavor)
-                            print_result('entropy-check-entropy-avail', check_entropy_avail(fconn, image.name))
-                            print_result('entropy-check-rngd', check_rngd(fconn, image.name))
-                            print_result('entropy-check-fips-test', check_fips_test(fconn, image.name))
+                        server = create_vm(conn, all_flavors, image)
+                        console = conn.compute.get_server_console_output(server)
+                        while True:
```
[the remark this answers is not in the document]
Comment (mbuechse) [reply Tests/iaas/entropy/entropy-check.py:455]: It's really hard to tell how long this takes, but let's say we won't accept any longer than 5 minutes.

[the remark this answers is not in the document]
Comment (mbuechse) [reply Tests/iaas/entropy/entropy-check.py]: 0c93ddc4343e093236ac14628fde931edf492183

**Review by garloff — APPROVED** [review]
I did some testing and it works.
(Some unrelated improvement: Only consider public images when determining image-properties. Will open a PR for this.)

**Comment by garloff** [comment]
Please consider merging #724.
Though this of course could also be done in a PR that goes directly against main, as it's not dependant ... as you prefer.

**Comment by berendt** [comment]
Independent of the PR: The default error messages should be improved:

```
❯ python3 entropy-check.py --os-cloud=regiocloud
entropy-check-image-properties: PASS
entropy-check-flavor-properties: PASS
CRITICAL: ConflictException('ConflictException: 409')
```

Real problem (visible with `--debug`): `openstack.exceptions.ConflictException: ConflictException: 409: Client Error for url: https://neutron.services.a.regiocloud.tech/v2.0/routers, Quota exceeded for resources: ['router'].`

Test itself looks good to me:

```
❯ python3 entropy-check.py --os-cloud=regiocloud --debug
DEBUG: Connecting to cloud 'regiocloud'
DEBUG: Selected image: Debian 12 (3ec7e315-d7d0-4de4-a4a2-2dd7473b0059)
DEBUG: Checking images and flavors for recommended attributes
entropy-check-image-properties: PASS
entropy-check-flavor-properties: PASS
DEBUG: Checking dynamic instance properties
DEBUG: Using external network e6be7364-bfd8-4de7-8120-8f41c69a139a.
DEBUG: Using userdata:
#cloud-config
apt:
  primary:
    - arches: [default]
      uri: https://mirror.plusserver.com/debian/debian/
  security: []
packages:
  - rng-tools5
runcmd:
  - echo '_scs-test-entropy-avail'; cat /proc/sys/kernel/random/entropy_avail
  - echo '_scs-test-fips-test'; cat /dev/random | rngtest -c 1000
  - echo '_scs-test-rngd'; sudo systemctl status rngd
  - echo '_scs-test-virtio-rng'; cat /sys/devices/virtual/misc/hw_random/rng_available; sudo /bin/sh -c 'od -vAn -N2 -tu2 < /dev/hwrng'
  - echo '_scs-test-end'
final_message: "_scs-test-end"
DEBUG: Creating instance of image 'Debian 12' using flavor 'SCS-1L-1' and 8 GiB ephemeral boot volume
DEBUG: Server '_scs-0101-server' ('335ece86-4327-44d6-91c0-dbc40d264f37') has been created
DEBUG: Finished waiting with timeout remainder: 69 s
entropy-check-entropy-avail: PASS
entropy-check-rngd: PASS
entropy-check-fips-test: PASS
DEBUG: Deleting server '_scs-0101-server'
DEBUG: Total critical / error / warning: 0 / 0 / 0
entropy-check: PASS
```

Not sure if this is a good idea. I think this is not necessary at all:

```
apt:
  primary:
    - arches: [default]
      uri: https://mirror.plusserver.com/debian/debian/
```

At least I see this issue:

```
[...]
[  166.346854] cloud-init[528]: Err:7 https://mirror.plusserver.com/debian/debian bookworm-security Release
[  166.350908] cloud-init[528]:   404  Not Found [IP: 62.138.222.4 443]
[...]
[  174.755329] cloud-init[528]: E: The repository 'mirror+file:/etc/apt/mirrors/debian-security.list bookworm-security Release' does not have a Release file.
[  174.765189] cloud-init[528]: 2024-08-30 19:36:05,330 - util.py[WARNING]: Package update failed
```

Also I think it is not necessary to assign a floating IP address to the instances.

**Comment by mbuechse** [comment]
@berendt Setting the repo is necessary because otherwise it will be set kinda randomly, which has led to Heisenbugs in the past.

What is up with the floating IP? The whole point of this change is to get rid of it. And we no longer set auto_ip=True?

**Comment by berendt** [comment]
I'll run it again tomorrow. I'm pretty sure that the instance had a floating IP address on it. I had cleared the test project, but maybe I was not looking at the right place.

With the repo, we at least had the error from above on REGIO.cloud. The error does not appear to prevent the test itself. In addition, by setting an external repository, the assumption is made that this is also accessible from the instance. We have at least one environment (not the FITKO POC) that is fuly airgapped. Such an external repository cannot be reached there. Especially with private clouds, you often only have the option of accessing local repository servers (hardcoded in custom cloud images).

**Comment by berendt** [comment]
Bildschirmfoto 2024-09-01 um 19 53 53

**Comment by berendt** [comment]
I tested with `auto_ip=False`:

```
    server = env.conn.create_server(
        server_name, image=image, flavor=flavor, userdata=userdata, wait=True, timeout=500,
        boot_from_volume=True, terminate_volume=True, volume_size=volume_size, network=env.network,
        auto_ip=False
    )
```

Looks better now:

Bildschirmfoto 2024-09-01 um 19 59 00

Test result looks good:

```
❯ python3 entropy-check.py --os-cloud=regiocloud --debug
DEBUG: Connecting to cloud 'regiocloud'
DEBUG: Selected image: Debian 12 (3ec7e315-d7d0-4de4-a4a2-2dd7473b0059)
DEBUG: Checking images and flavors for recommended attributes
entropy-check-image-properties: PASS
entropy-check-flavor-properties: PASS
DEBUG: Checking dynamic instance properties
DEBUG: Using external network e6be7364-bfd8-4de7-8120-8f41c69a139a.
DEBUG: Using userdata:
#cloud-config
apt:
  primary:
    - arches: [default]
      uri: https://mirror.plusserver.com/debian/debian/
  security: []
packages:
  - rng-tools5
runcmd:
  - echo '_scs-test-entropy-avail'; cat /proc/sys/kernel/random/entropy_avail
  - echo '_scs-test-fips-test'; cat /dev/random | rngtest -c 1000
  - echo '_scs-test-rngd'; sudo systemctl status rngd
  - echo '_scs-test-virtio-rng'; cat /sys/devices/virtual/misc/hw_random/rng_available; sudo /bin/sh -c 'od -vAn -N2 -tu2 < /dev/hwrng'
  - echo '_scs-test-end'
final_message: "_scs-test-end"
DEBUG: Creating instance of image 'Debian 12' using flavor 'SCS-1L-1' and 8 GiB ephemeral boot volume
DEBUG: Server '_scs-0101-server' ('ba77a314-eabe-4d03-a3c2-1000ffecf5cf') has been created
DEBUG: Finished waiting with timeout remainder: 278 s
entropy-check-entropy-avail: PASS
entropy-check-rngd: PASS
entropy-check-fips-test: PASS
DEBUG: Deleting server '_scs-0101-server'
DEBUG: Total critical / error / warning: 0 / 0 / 0
entropy-check: PASS
```

**Review by martinmo — COMMENTED** [review]
I did a further round of testing and inspired by @berendt I checked the console log and the IPs of the `_scs-0101-server` that's created during the test. Observations:

* Can confirm that a Floating IP is still attached unless `auto_ip=False` is passed to the `create_server(…)` call, this needs to be changed
* Can confirm the 404 for the plusserver mirror
* There might be some problems with the cloud-init configuration – I saw this in the logs:

```
...
[   62.091344] cloud-init[394]: 2024-09-02 07:46:19,174 - schema.py[WARNING]: Invalid cloud-config provided: Please run 'sudo cloud-init schema --system' to see the schema errors.
...
```

**Review by martinmo — APPROVED** [review]

**Pull request merged**